lightning/src/offers/static_invoice.rs

```diff
@@ -0,0 +1,1173 @@
+				return Err(Bolt12SemanticError::AlreadyExpired);
+			}
+		}
+
+		let Self { offer_bytes, invoice, keys } = self;
+		let unsigned_invoice = UnsignedStaticInvoice::new(&offer_bytes, invoice);
+		let invoice = unsigned_invoice
+			.sign(|message: &UnsignedStaticInvoice| {
+				Ok(secp_ctx.sign_schnorr_no_aux_rand(message.tagged_hash.as_digest(), &keys))
+			})
+			.unwrap();
+		Ok(invoice)
+	}
+
+	invoice_builder_methods_common!(self, Self, self.invoice, Self, self, S, StaticInvoice, mut);
+}
+
+/// A semantically valid [`StaticInvoice`] that hasn't been signed.
+pub struct UnsignedStaticInvoice {
+	bytes: Vec<u8>,
+	contents: InvoiceContents,
+	tagged_hash: TaggedHash,
+}
+
+macro_rules! invoice_accessors { ($self: ident, $contents: expr) => {
+	/// The chain that must be used when paying the invoice. [`StaticInvoice`]s currently can only be
+	/// created from offers that support a single chain.
+	pub fn chain(&$self) -> ChainHash {
+		$contents.chain()
+	}
+
+	/// Opaque bytes set by the originating [`Offer::metadata`].
+	///
+	/// [`Offer::metadata`]: crate::offers::offer::Offer::metadata
+	pub fn metadata(&$self) -> Option<&Vec<u8>> {
+		$contents.metadata()
+	}
+
+	/// The minimum amount required for a successful payment of a single item.
+	///
+	/// From [`Offer::amount`].
+	///
+	/// [`Offer::amount`]: crate::offers::offer::Offer::amount
+	pub fn amount(&$self) -> Option<Amount> {
+		$contents.amount()
+	}
+
+	/// Features pertaining to the originating [`Offer`], from [`Offer::offer_features`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	/// [`Offer::offer_features`]: crate::offers::offer::Offer::offer_features
+	pub fn offer_features(&$self) -> &OfferFeatures {
+		$contents.offer_features()
+	}
+
+	/// A complete description of the purpose of the originating offer, from [`Offer::description`].
+	///
+	/// [`Offer::description`]: crate::offers::offer::Offer::description
+	pub fn description(&$self) -> Option<PrintableString> {
+		$contents.description()
+	}
+
+	/// Duration since the Unix epoch when an invoice should no longer be requested, from
+	/// [`Offer::absolute_expiry`].
+	///
+	/// [`Offer::absolute_expiry`]: crate::offers::offer::Offer::absolute_expiry
+	pub fn absolute_expiry(&$self) -> Option<Duration> {
+		$contents.absolute_expiry()
+	}
+
+	/// The issuer of the offer, from [`Offer::issuer`].
+	///
+	/// [`Offer::issuer`]: crate::offers::offer::Offer::issuer
+	pub fn issuer(&$self) -> Option<PrintableString> {
+		$contents.issuer()
+	}
+
+	/// Paths to the node that may supply the invoice on the recipient's behalf, originating from
+	/// publicly reachable nodes. Taken from [`Offer::paths`].
+	///
+	/// [`Offer::paths`]: crate::offers::offer::Offer::paths
+	pub fn offer_message_paths(&$self) -> &[BlindedPath] {
+		$contents.offer_message_paths()
+	}
+
+	/// Paths to the recipient for indicating that a held HTLC is available to claim when they next
```

Comment:
For refreshing my understanding, who is holding the HTLC here? Is this for sending `held_htlc_available`? Do you have a reference for this flow? I didn't see it in the spec PR other than:

>   - if `invoice_payment_hash` is unset:
>     - MUST reject the invoice if `invoice_message_paths` is not present or is empty.
>     - MUST pay asynchronously using the `held_htlc_available` onion message
>       flow, where the onion message is sent over `invoice_message_paths`.

Reply:
> who is holding the HTLC here?

Either the always-online sender or the often-offline sender's LSP. 

> Is this for sending held_htlc_available?

Yep!

For reference, https://github.com/lightning/bolts/pull/1149 is based on https://github.com/lightning/bolts/pull/989, which is the flow referred to in the quote. Let me know if something's unclear.

Reply:
Ah, thanks! I was only looking at the last commits and not also the first ones. 🤦‍♂️